source/develop/release-management/features/virt/parallel-migration-connections.md

```diff
@@ -0,0 +1,88 @@
+
+Additionally, multifd implementation of data transfer is simpler and more efficient than the traditional migration mechanism and can be faster even with lower network bandwidths.  This concerns only the way memory data is transferred, the other migration concepts such as counting iterations or setting the maximum downtime are unaffected.
+
+Parallel migration connections can be enabled by passing ``VIR_MIGRATE_PARALLEL`` flag to [virDomainMigrateToURI3](https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainMigrateToURI3) call.  The number of connections to use is specified using  [VIR_MIGRATE_PARAM_PARALLEL_CONNECTIONS](https://libvirt.org/html/libvirt-libvirt-domain.html#VIR_MIGRATE_PARAM_PARALLEL_CONNECTIONS) migration parameter.
+
+## Prerequisites
+
+Nothing special, just up-to-date libvirt and QEMU.
+
+## Limitations
+
+Current implementation of parallel migration connections in QEMU has some restrictions:
+
+- It doesn't support switching to post-copy.
+
+- RDMA doesn't work.  This is not a limitation for oVirt, because RDMA is not supported in oVirt.
+
+- XBZRLE compression is not supported.  However this kind of compression is generally not recommended to use in migrations because it adds little benefit and a significant overhead.  gzip/zstd is supported instead and migrations should actually benefit from it.  (TODO: Check whether libvirt supports this kind of compression.)
+
+- Due to their better implementation, parallel migration connections can be faster than the traditional migration implementation even when the network bandwidth can be saturated with a single connection.  However, when single connection is specified, the parallel migration connection implementation uses a different code path, which may not be sufficiently robust, and it's recommended to always use at least two connections.  In the worst case, one of the threads will be idle while causing only a little overhead. 
+
+- Parallel migration connections haven't been tested extensively with connections faster than 100 Gbps.  Due to current multifd implementation, there is an increasing overhead when using many connections and it's not clear whether using more than 16 parallel connections adds a real benefit.
+
+- It still applies that large guests with terrabytes of RAM are difficult to migrate successfully.
+
+- Since it's easier to saturate network bandwidth with multiple connections, it's advisable to not migrate multiple VMs from a single host at the same time.
+
+- VM CPU pinning may restrict the number of CPUs available to the QEMU process and thus impose a limit on the meaningful number of parallel connections.
```

Comment:
It's not necessarily bad if pinned/dedicated CPUs are used for migration threads. This helps to speed up the migration and, unlike autoconverge applied in this case, it doesn't waste CPU capacity. However I'm not sure whether/what means we have to influence the placement of the migration threads, I must look at it.

Reply:
ack

Reply:
Just to clarify, will the intention be to use the VM CPUs instead of reserved or shared CPUs? The would certainly speed the convergence up but at the expense of degraded performance of the migrated VM during the process. Bear in mind that we are talking about VMs tuned for performance. Still, maybe that is something users are willing to accept during migration?

A consequence of this would be that you are also limited by number of CPUs of the VM in cases it has less than 16. An option would be to use combination of VMs CPUs and a shared pool. But using a shared pool complicates things quite a bit because we need to adapt the VM's configuration every time the shared pool changes. That being said...

Of course the above is not directly applicable to VMs without policy that run on a shared pool. And we probably should not allow migration of VM to affect performance of other VMs with dedicated CPUs (or pinning). That means VMs running in shared pool should use only CPUs in shared pool for migration. So it seems we wont be spared of adapting to shared pool changes during migration anyway.

It seems to me like a can of worms. It should be checked what is the real CPU consumption of the migration threads -- maybe all this is not needed and it would be best to leave the CPU management to the host instead of trying to manage this.

Reply:
> Just to clarify, will the intention be to use the VM CPUs instead of reserved or shared CPUs?

That's a good question, yeah, we need to see what means we have to influence the placement of the migration threads as Milan commented above and then see what we can do - reading my comment above, I thought about the latter (using reserved/shared CPUs) but that's not necessarily right (as you wrote) or possible.

> A consequence of this would be that you are also limited by number of CPUs of the VM in cases it has less than 16.

Yeah, this is taken into account in [calculateAutomaticParallelMigrations](https://gerrit.ovirt.org/c/ovirt-engine/+/116775/15/backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/MigrateVmCommand.java#563)

Reply:
My current intention is to leave the selection of CPUs to be used for the migration threads to the host. We don't want to pin the emulator threads (which I think are used for migrations) just because of migrations (a different thing would be if e.g. VMs with dedicated CPUs wanted to pin the emulator threads, but even in such a case this would just limit the maximum number of meaningful migration connections) and to complicate things without a good reason.

As for running migration threads on VM CPUs: If the VM wants to prevent that (e.g. because of dedicated CPUs), it should arrange the emulator thread pinning itself. It's true that migration threads, which can consume a lot of CPU power if the right number of threads is selected, can slow down dedicated CPUs if they land on them. But take into account that all our current migration policies use autoconverge, which slows down VM CPUs artificially if the migration doesn't converge. The difference is that migration threads can slow down CPUs since the beginning of the migration. But unless we want to disable autoconverge, which might prevent some VMs from finishing their migrations, VMs cannot expect having full CPU power during migrations even when they have dedicated or pinned CPUs.

Yes, another problem is if migration threads use CPUs dedicated to a different VM than the migrating one. We have two reasonable options here: to rely on the OS scheduling to do something reasonable or to manage emulator thread pinning. But again, if we want to do the latter then it should be a part of the dedicated CPUs functionality, since other QEMU stuff can possibly run on the dedicated CPUs, including current migrations.

Reply:
With the addition to the dedicated CPU page, I think that's fair enough